etl/src/main/java/org/pih/hivmigration/etl/sql/UnknownEncounterLocationsMigrator.groovy

```diff
@@ -0,0 +1,214 @@
+            where
+                e.patient_id = pp.patient_id and e.encounter_datetime < pp.date_enrolled
+                and e.location_id = @unknown_location
+                and pp.program_id = @hiv_program
+            group by 
+                e.encounter_id;
+    
+            # create temp table to map most recent program location to encounter
+            drop temporary table if exists  tmp_most_recent_subsequent_program_location_to_encounter;
+            create temporary table tmp_most_recent_subsequent_program_location_to_encounter (patient_id int, encounter_id int, encounter_datetime datetime, location_id int);
+    
+            # populate most recent program location tmp table by joining encounter and patient_program using most_recent_subsequent_program_date_to_encounter temp table
+            insert into tmp_most_recent_subsequent_program_location_to_encounter
+                select 
+                    e.patient_id, e.encounter_id, e.encounter_datetime, pp.location_id
+                from 
+                    tmp_most_recent_subsequent_program_date_to_encounter tmp, patient_program pp, encounter e
+                where 
+                    e.encounter_id = tmp.encounter_id and pp.date_enrolled = tmp.date_enrolled and pp.patient_id = e.patient_id;
+        
+            # do the actual update
+            update encounter e, tmp_most_recent_subsequent_program_location_to_encounter tmp
+                set e.location_id = tmp.location_id
+                where e.encounter_id = tmp.encounter_id;
+    
+            # log a data warning
+            insert into hivmigration_data_warnings (openmrs_patient_id, openmrs_encounter_id, warning_details, encounter_date, flag_for_review, warning_type)
+                select 
+                    patient_id as openmrs_patient_id, encounter_id as openmrs_encounter_id,  CONCAT('Updated to location :', location_id) as warning_details, 
+                    DATE(encounter_datetime) as encounter_date, 0 as flag_for_review, 
+                    'Encounter location changed from Unknown to location of Most Recent Subsequent Program Enrollment' as warning_type
+               from tmp_most_recent_subsequent_program_location_to_encounter;
+        ''');
+
+        executeMysql("Update any unknown visit locations that we now can resolve with the additional encounter information",
+'''
+            set @unknown_location = (select location_id from location where uuid='8d6c993e-c2cc-11de-8d13-0010c6dffd0f');
+            
+            # create temporary table to store new visit location
+            drop table if exists tmp_new_visit_location;
+            create temporary table tmp_new_visit_location  (visit_id int, location_id int);
+    
+            # find all visits that have a unknown location but also have or more encounters with a location other than Unknown Location
+            # pull the location off the matching visits (arbitrarily picking the "max" if multiple possibilities found)
```

Comment:
Do we need to think about this arbitrary value?  i.e. are some encounters more meaningful than others?  if there are 10 encounters at location X, and 1 at location Y, I'd think we'd want to use location X.  if there is one RDV encounter, and one Lab Results encounter or Drug Order Documentation encounter, I'd think we'd want to use the RDV encounter.